runtime/major_gc.c

```diff
@@ -167,26 +292,58 @@ void caml_darken (value v, value *p /* not used */)
     CAMLassert (!Is_blue_hd (h));
     if (Is_white_hd (h)){
       ephe_list_pure = 0;
+      Hd_val (v) = Blackhd_hd (h);
       if (t < No_scan_tag){
-        Hd_val (v) = Grayhd_hd (h);
-        *gray_vals_cur++ = v;
-        if (gray_vals_cur >= gray_vals_end) realloc_gray_vals ();
-      }else{
-        Hd_val (v) = Blackhd_hd (h);
+        mark_entry me = {v, 0};
+        mark_stack_push(Caml_state->mark_stack, me, NULL);
```

Comment:
Another reason not to pass structs to functions...  It would read better as
```
mark_stack_push(Caml_state->mark_stack, v, 0, NULL);
```

Reply:
Agree on this, I've changed `mark_stack_push` to take the components instead.